app/webpacker/controllers/productsV3_controller.js

```diff
@@ -0,0 +1,24 @@
+import ApplicationController from "./application_controller";
+
+export default class extends ApplicationController {
+  static targets = ["loading"];
+
+  connect() {
+    super.connect();
+    // Fetch the products on page load
+    this.load();
+  }
+
+  load = () => {
+    this.showLoading();
+    this.stimulate("Admin::ProductsV3#fetch").then(() => this.hideLoading());
+  };
```

Comment:
Maybe it's too early for extraction but I would like to see a more general solution for the "loading spinner. I would expect it to work similar to the `disable-with` attribute on submit buttons. Is there something pre-existing?

I don't know if it would work but we could have a controller inserting the spinner into the the target element when an action is triggered and then it's replaced with new HTML when the target element is re-rendered.

Reply:
The controller has changed since, don't know if it fits well for you.

It consists of a `loading-controller`, that works with event listener. 

```js
import ApplicationController from "./application_controller";

export default class extends ApplicationController {
  connect() {
    super.connect();
    document.addEventListener("show-loading", this.showLoading);
    document.addEventListener("hide-loading", this.hideLoading);
  }

  disconnect() {
    document.removeEventListener("show-loading", this.showLoading);
    document.removeEventListener("hide-loading", this.hideLoading);
  }

  hideLoading = () => {
    this.element.classList.add("hidden");
  };

  showLoading = () => {
    this.element.classList.remove("hidden");
  };
}
```

and then `products-controller` is simply:

```js
import ApplicationController from "./application_controller";

export default class extends ApplicationController {
  connect() {
    super.connect();
    // Fetch the products on page load
    this.stimulate("Products#fetch");
  }

  beforeReflex() {
    this.showLoading();
  }

  afterReflex() {
    this.hideLoading();
  }

  showLoading = () => {
    const event = new CustomEvent("show-loading");
    document.dispatchEvent(event);
  };

  hideLoading = () => {
    const event = new CustomEvent("hide-loading");
    document.dispatchEvent(event);
  };
}
```

Reply:
Actually, writing my comment, I thought I could do in another (smarter?) way ; See commit 9e5078d59c

Reply:
That's an interesting way of doing it. I don't have an opinion about it. :shrug: